library/src/scala/Tuple.scala

```diff
@@ -338,9 +338,15 @@ object Tuple:
      */
     transparent inline def indexOf(y: Any): Int = constValue[IndexOf[X, y.type]]
 
-    /** A boolean indicating whether there is an element `y.type` in the type `X` of `x`
-     */
-    transparent inline def contains(y: Any): Boolean = constValue[Contains[X, y.type]]
+    /** A boolean indicating whether there is an element `y.type` in the type `X` of `x` */
```

Comment:
> Equality could mix values with different types

I'm not sure I understood, did you mean the types in `X` could be widened ?

My reasoning, perhaps incorrect, was the following. For the definition to be sound, we only need to verify that the term and type level agree in the cases where the match type reduces.

If `x` is an empty tuple, then the result is immediate.
Otherwise let `X := x1 *: xs1` and `x := (hd: x1) *: (tl: xs1)`, we need to check 
1. `Contains[X, y.type]` reduces the the first case (i.e. true), implies `x1` matches `y.type`, implies `hd == y`
2. `Contains[X, y.type]` proceeds to the second case, implies `x1` is disjoint from `y.type`, implies `hd != y`

Which I believe is correct according to the match type spec. \cc @sjrd

Reply:
Not directly linked, but interesting; we can implement an unoptimised version which does not require using `asInstanceOf`. Provided restructuring of the `type Contains` to account for limitations of gadt reasoning.

```scala 3
type Contains1[X <: Tuple, Y] <: Boolean = X match
  case x *: xs => x match
    case Y => true
    case _ => Contains1[xs, Y]
  case EmptyTuple => false

extension [X <: Tuple](self: X) def contains1(y: Any): Contains1[X, y.type] = 
  self match
    case self: (x *: xs) =>
      val (hd: x) *: (tl: xs) = self
      hd match
        case _: y.type => true
        case _ => tl.contains1(y)
    case _: EmptyTuple => false

val tup = (1, 2, 3)
assert(tup.contains1(2))
assert(!tup.contains1(0))
assert(!tup.contains1("hi"))
```

Reply:
> Equality could mix values with different types

I think it means you can have a value in the tuple that always equals whatever is passed to it (e.g. `a`), so `a.type` the singleton reference `a` might not be in the tuple.

Reply:
> Contains[X, y.type] proceeds to the second case, implies x1 is disjoint from y.type, implies hd != y

Not necessarily. For instance `Int` is disjoint from `Float` yet an `Int value can `==` a `Float` value.

Reply:
Indeed, good catch ! I did not think of that.

We're coming back to the original version, but putting this here for reference in case we ever reconsider the tuple definitions. The issue can by fixed by replacing `productIterator` with  `toArray` which returns objects and requiring `y: AnyRef`.